### PR TITLE
qwen3-14b: make the fused decode_fwd use paged attention

### PR DESCRIPTION
**Pull request by lwDavid**
## Summary

Make the fused Qwen3-14B decode (`decode_fwd` / `_decode_layer` in `models/qwen3/14b/decode_layer.py`) use **PAGED attention**, so it consumes the serving `block_table` + `slot_mapping` and reads/writes the SAME paged KV pool `prefill_fwd` writes — instead of the non-paged contiguous `[layer, batch, kv_head, seq]` cache. This lets the serving decode drop its paged→contiguous bridge and work under multi-request continuous batching (KV is addressed by `block_table`, not kernel row).

## Changes

- `decode_fwd` / `_decode_layer` / `decode_fwd_layers` / `qwen3_decode_mpmd` take `block_table` + `slot_mapping` (dynamic dims); `k_cache` / `v_cache` become runtime-dynamic (`KV_CACHE_ROWS_DYN`). Per-layer cache stride + `max_blocks_per_seq` are derived from the tensor dims, like `prefill_fwd`.
- `SEQ_TILE 256 → 128` to match the page size, so one logical KV block is exactly one physical page (one contiguous per-block slice). `BLOCK_SIZE = page_size = 128`.
- Paged **write** in `rope_qkv` (`slot_mapping` → `(slot_block*NUM_KV_HEADS + ki)*BLOCK_SIZE + offset`) and paged **read** in `fa_fused` (`block_table[b*max_blocks + sb]` → pbid → `(pbid*NUM_KV_HEADS + kvh)*BLOCK_SIZE`). The dense block work-table, the layer loop and the LM-head fusion are unchanged.
- `_smoke_inputs` / `INPUT_NAMES` / `--validate-fwd` updated to the paged signature.

## Testing

- `python decode_layer.py -p a2a3sim --smoke` compiles end-to-end (the data-dependent `block_table` / `slot_mapping` reads codegen OK).
- Wired into pypto-serving and run e2e on a2a3 (prompt "Huawei is") → identical coherent 32 tokens, decode ~8.8 tok/s.

## Notes

Still fixed `BATCH=16` / fixed `NUM_LAYERS=40` (kernel constants). Pairs with the pypto-serving change that dispatches it (passes `block_table` / `slot_mapping` + the shared paged pool and removes the contiguous-KV bridge); the serving submodule must bump to this commit.

### CODE REVIEW

File: models/qwen3/14b/decode_layer.py [hunk 1]
```diff
@@ -64,7 +64,12 @@
 from pypto.backend import BackendType, set_backend_type
 from pypto.runtime import RunConfig
 
-from config import VOCAB  # vocab size for the fused decode_fwd LM head / logits
+from config import (  # shared model/serving config
+    BLOCK_TABLE_FLAT_DYN,  # flat [user_batch * max_blocks_per_seq] block table dim
+    KV_CACHE_ROWS_DYN,     # runtime-dynamic paged KV pool row count
+    USER_BATCH_DYN,        # runtime-dynamic user batch
+    VOCAB,                 # vocab size for the fused decode_fwd LM head / logits
+)
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/qwen3/14b/decode_layer.py]: Fixed in `09ec0ca` — removed the three unused dynamic-dimension imports (`BLOCK_TABLE_FLAT_DYN`, `KV_CACHE_ROWS_DYN`, `USER_BATCH_DYN`). The decode path derives these dims at runtime via `pl.tensor.dim(...)`, so the markers were dead imports. ruff-check now passes.

File: models/qwen3/14b/decode_layer.py [hunk 2]
```diff
@@ -264,6 +277,8 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
     q_norm_weight: pl.Tensor,
     k_norm_weight: pl.Tensor,
     seq_lens: pl.Tensor,
+    block_table: pl.Tensor,
+    slot_mapping: pl.Tensor,
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/qwen3/14b/decode_layer.py:276]: Skipping: this file deliberately uses bare `pl.Tensor` annotations on every host entry (`qwen3_decode_mpmd`, `decode_fwd`, `decode_fwd_layers`) and derives all dynamic dims at runtime via `pl.tensor.dim(...)` — it never binds the DYN markers (unlike `prefill_fwd.py`). Annotating only the `@pl.jit.inline` helper would bind nothing and diverge from the file's design. The imports have been removed instead (`09ec0ca`).

File: models/qwen3/14b/decode_layer.py [hunk 3]
```diff
@@ -276,12 +291,19 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
...
     layer_inter_base = layer_idx * INTERMEDIATE
-    layer_cache_base = layer_idx * (BATCH * NUM_KV_HEADS * MAX_SEQ)
+    # Paged KV: rows are runtime-dynamic (the paged pool sizes them). Derive the
+    # per-layer stride and the block-table row stride from the tensor dims, exactly
+    # as prefill_fwd does, so decode reads the SAME pool prefill wrote.
+    num_layers_actual = pl.tensor.dim(input_rms_weight, 0)
+    layer_cache_rows = pl.tensor.dim(k_cache, 0) // num_layers_actual
+    layer_cache_base = layer_idx * layer_cache_rows
+    user_batch = pl.tensor.dim(seq_lens, 0)
+    max_blocks_per_seq = pl.tensor.dim(block_table, 0) // user_batch
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/qwen3/14b/decode_layer.py:301]: Skipping: the runtime derivation is intentional. As the surrounding comment states, decode derives `user_batch`/`max_blocks_per_seq` from the tensor dims "exactly as prefill_fwd does, so decode reads the SAME pool prefill wrote." Hardcoding compile-time constants would break that paged-pool consistency with prefill. `BATCH` is fixed at 16 but `user_batch` (ragged request count) and the paged row count are genuinely runtime-dynamic.

File: models/qwen3/14b/decode_layer.py [hunk 4]
```diff
@@ -627,9 +654,13 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
...
             fa_ctx_len = pl.read(seq_lens, [fa_b])
             # Table holds only real blocks → exactly one block per entry, at fa_p.
-            sb = fa_p  # global KV block index (no inner loop — old p_blocks was 1)
+            sb = fa_p  # logical KV block index (no inner loop — old p_blocks was 1)
             s0 = sb * SEQ_TILE
             valid_len = pl.min(SEQ_TILE, fa_ctx_len - s0)
+            # Paged read: map logical block sb -> physical page via this request's
+            # block_table row. SEQ_TILE == page_size, so one page is exactly one
+            # contiguous SEQ_TILE-row slice of the pool (shared with the kvh below).
+            fa_pbid = pl.cast(pl.tensor.read(block_table, [fa_b * max_blocks_per_seq + sb]), pl.INDEX)
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/qwen3/14b/decode_layer.py:658]: Skipping for the same reason as the `max_blocks_per_seq` derivation above: the block-table row stride is derived at runtime to stay consistent with the paged pool prefill wrote. Replacing it with the compile-time `MAX_CTX_BLOCKS` would couple block indexing to a constant that may not match the runtime block-table layout.
